Example15813.hs

```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE DamlSyntax #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+module Example15813 where
+
+import DA.Internal.Desugar
+import GHC.Types
+
+template T
+  with
+    p : Party
+  where
+    signatory p
+
+    choice C : ()
+      controller p
+      do pure ()
+
+template U
+  with
+    p2 : Party
+  where
+    signatory p2
+
+abort : Text -> Update a
+abort = undefined
+
+exerciseE : Choice t c (Either Text r) => ContractId t -> c -> Update r
+exerciseE cid c = do
+  r <- exercise cid c
+  case r of
+    Left err -> abort err
+    Right r -> pure r
+
+test : ContractId T -> Update ()
+test cid = exerciseE cid C
```

Comment:
when I run `./_build/stage1/bin/ghc -fforce-recomp Example15813.hs` I get the plain GHC error in terms of constraints:

```
[1 of 3] Compiling GHC.Types        ( GHC/Types.hs, GHC/Types.o )
[2 of 3] Compiling DA.Internal.Desugar ( DA/Internal/Desugar.hs, DA/Internal/Desugar.o )
[3 of 3] Compiling Example15813     ( Example15813.hs, Example15813.o )

Example15813.hs:42:12: error:
    • Couldn't match type ‘()’ with ‘Either Text ()’
        arising from a functional dependency between:
          constraint ‘HasExercise T C (Either Text ())’
            arising from a use of ‘exerciseE’
          instance ‘HasExercise T C ()’ at <no location info>
    • In the expression: exerciseE cid C
      In an equation for ‘test’: test cid = exerciseE cid C
   |
42 | test cid = exerciseE cid C
   |           
   ```